File: Services/ScreenTargeting.cs
using System.Diagnostics;
using System.Drawing;
using Forms = System.Windows.Forms;

namespace PrecisionJump.Services;

public sealed record DisplayMonitor(
    string DeviceName,
    Rectangle Bounds,
    Rectangle WorkingArea,
    bool IsPrimary,
    int Number);

public sealed record NineGridPreview(
    DisplayMonitor Display,
    Point ActualCursor,
    int Depth,
    double MapScale,
    double StepX,
    double StepY)
{
    public bool IsScreenLevel => Depth == 0;

    public string Label => IsScreenLevel
        ? $"SCREEN {Display.Number}"
        : $"ZOOM {MapScale:F1}×";
}

public sealed record GridInteraction(
    NineGridPreview Preview,
    Point? JumpTarget);

public sealed class NineGridSession
{
    private const double ScaleAnimationSeconds = 0.14;

    private readonly IReadOnlyList<DisplayMonitor> _displays;
    private double _mapX;
    private double _mapY;
    private int _depth;
    private double _scale = 1;
    private double _scaleStart = 1;
    private double _scaleTarget = 1;
    private long _scaleStartedAt;

    public NineGridSession(
        Point origin,
        IReadOnlyList<DisplayMonitor> displays)
    {
        if (displays.Count == 0)
        {
            throw new InvalidOperationException("No displays are available.");
        }

        _displays = displays;
        Origin = origin;
        _mapX = origin.X;
        _mapY = origin.Y;
        Display = FindDisplay(origin);
        Preview = CreatePreview();
    }

    public Point Origin { get; }
    public Point CurrentPosition => RoundedPosition();
    public DisplayMonitor Display { get; private set; }
    public int Depth => _depth;
    public double MapScale => _scale;
    public double StepX => Display.Bounds.Width / Math.Max(_scale, 1);
    public double StepY => Display.Bounds.Height / Math.Max(_scale, 1);
    public NineGridPreview Preview { get; private set; }

    public bool IsScaleAnimating
    {
        get
        {
            var elapsed = (Stopwatch.GetTimestamp() - _scaleStartedAt)
                / (double)Stopwatch.Frequency;
            return Math.Abs(_scale - _scaleTarget) > 0.001
                && elapsed < ScaleAnimationSeconds;
        }
    }

    public GridInteraction MoveContinuous(
        double physicalDeltaX,
        double physicalDeltaY,
        double mapUnitTravelDistance,
        long? nowTicks = null)
    {
        UpdateScale(nowTicks);
        if (physicalDeltaX == 0 && physicalDeltaY == 0)
        {
            Preview = CreatePreview();
            return new GridInteraction(Preview, null);
        }

        var travel = Math.Max(mapUnitTravelDistance, 1);
        MoveOnAdjacentDisplayMap(
            physicalDeltaX / (travel * _scale),
            physicalDeltaY / (travel * _scale));

        Preview = CreatePreview();
        return new GridInteraction(Preview, CurrentPosition);
    }

    public GridInteraction ZoomIn(
        int maximumDepth,
        Point? actualPosition = null,
        long? nowTicks = null)
    {
        if (Depth >= maximumDepth)
        {
            return new GridInteraction(Preview, null);
        }

        if (actualPosition is Point position)
        {
            SyncPosition(position);
        }
        StartScaleTransition(Depth + 1, nowTicks);
        Preview = CreatePreview();
        return new GridInteraction(Preview, null);
    }

    public GridInteraction ZoomOut(
        Point? actualPosition = null,
        long? nowTicks = null)
    {
        if (Depth == 0)
        {
            return new GridInteraction(Preview, null);
        }

        if (actualPosition is Point position)
        {
            SyncPosition(position);
        }
        StartScaleTransition(Depth - 1, nowTicks);
        Preview = CreatePreview();
        return new GridInteraction(Preview, null);
    }

    public NineGridPreview RefreshScale(long? nowTicks = null)
    {
        UpdateScale(nowTicks);
        Preview = CreatePreview();
        return Preview;
    }

    public void AcceptProgrammaticPosition(Point actualPosition)
    {
        SyncPosition(actualPosition);
        Preview = CreatePreview();
    }

    private void MoveOnAdjacentDisplayMap(
        double normalizedDeltaX,
        double normalizedDeltaY)
    {
        var bounds = Display.Bounds;
        var width = Math.Max(bounds.Width - 1, 1);
        var height = Math.Max(bounds.Height - 1, 1);
        var u = Math.Clamp((_mapX - bounds.Left) / width, 0, 1);
        var v = Math.Clamp((_mapY - bounds.Top) / height, 0, 1);
        var remainingX = normalizedDeltaX;
        var remainingY = normalizedDeltaY;

        var maximumTransitions = Math.Max(_displays.Count * 2, 4);
        for (var iteration = 0; iteration < maximumTransitions; iteration++)
        {
            if (Math.Abs(remainingX) < 0.000001
                && Math.Abs(remainingY) < 0.000001)
            {
                break;
            }

            var horizontalTime = BoundaryTime(u, remainingX);
            var verticalTime = BoundaryTime(v, remainingY);
            var travelTime = Math.Min(1, Math.Min(horizontalTime, verticalTime));
            u += remainingX * travelTime;
            v += remainingY * travelTime;
            if (travelTime >= 1)
            {
                break;
            }

            var remainder = 1 - travelTime;
            remainingX *= remainder;
            remainingY *= remainder;

            // Resolve one shared edge at a time. A diagonal movement therefore
            // still passes through the screen that owns the first crossed edge.
            if (horizontalTime <= verticalTime)
            {
                var direction = remainingX > 0 ? 1 : -1;
                var crossingY = bounds.Top
                    + Math.Clamp(v, 0, 1) * height;
                var nextDisplay = FindHorizontalNeighbor(
                    direction,
                    crossingY);
                if (nextDisplay is null)
                {
                    remainingX = 0;
                    u = Math.Clamp(u, 0, 1);
                    continue;
                }

                Display = nextDisplay;
                bounds = nextDisplay.Bounds;
                width = Math.Max(bounds.Width - 1, 1);
                height = Math.Max(bounds.Height - 1, 1);
                u = direction > 0 ? 0 : 1;
                // Once two displays share a side, treat their entire logical
                // edges as aligned. Keeping the normalized perpendicular
                // position removes dead zones caused by size or offset.
                v = Math.Clamp(v, 0, 1);
            }
            else
            {
                var direction = remainingY > 0 ? 1 : -1;
                var crossingX = bounds.Left
                    + Math.Clamp(u, 0, 1) * width;
                var nextDisplay = FindVerticalNeighbor(
                    direction,
                    crossingX);
                if (nextDisplay is null)
                {
                    remainingY = 0;
                    v = Math.Clamp(v, 0, 1);
                    continue;
                }

                Display = nextDisplay;
                bounds = nextDisplay.Bounds;
                width = Math.Max(bounds.Width - 1, 1);
                height = Math.Max(bounds.Height - 1, 1);
                u = Math.Clamp(u, 0, 1);
                v = direction > 0 ? 0 : 1;
            }
        }

        _mapX = bounds.Left
            + Math.Clamp(u, 0, 1) * Math.Max(bounds.Width - 1, 1);
        _mapY = bounds.Top
            + Math.Clamp(v, 0, 1) * Math.Max(bounds.Height - 1, 1);
    }

    private static double BoundaryTime(double position, double delta)
    {
        if (delta > 0.000001)
        {
            return Math.Max(0, (1 - position) / delta);
        }

        if (delta < -0.000001)
        {
            return Math.Max(0, -position / delta);
        }

        return double.PositiveInfinity;
    }

    private void StartScaleTransition(int depth, long? nowTicks)
    {
        var now = nowTicks ?? Stopwatch.GetTimestamp();
        UpdateScale(now);
        _depth = depth;
        _scaleStart = _scale;
        _scaleTarget = Math.Pow(3, depth);
        _scaleStartedAt = now;
    }

    private void UpdateScale(long? nowTicks)
    {
        if (Math.Abs(_scale - _scaleTarget) <= 0.001)
        {
            _scale = _scaleTarget;
            return;
        }

        var now = nowTicks ?? Stopwatch.GetTimestamp();
        var elapsed = (now - _scaleStartedAt)
            / (double)Stopwatch.Frequency;
        var progress = Math.Clamp(
            elapsed / ScaleAnimationSeconds,
            0,
            1);
        var eased = progress * progress * (3 - 2 * progress);
        _scale = _scaleStart
            + (_scaleTarget - _scaleStart) * eased;
        if (progress >= 1)
        {
            _scale = _scaleTarget;
        }
    }

    private void SyncPosition(Point position)
    {
        _mapX = position.X;
        _mapY = position.Y;
        Display = FindDisplay(position);
    }

    private NineGridPreview CreatePreview()
    {
        return new NineGridPreview(
            Display,
            CurrentPosition,
            Depth,
            _scale,
            StepX,
            StepY);
    }

    private DisplayMonitor? FindHorizontalNeighbor(
        int direction,
        double crossingY)
    {
        var current = Display.Bounds;
        return _displays
            .Where(candidate => candidate.DeviceName != Display.DeviceName)
            .Where(candidate =>
                direction > 0
                    ? candidate.Bounds.Left == current.Right
                    : candidate.Bounds.Right == current.Left)
            .Where(candidate =>
                Overlaps(current.Top, current.Bottom,
                    candidate.Bounds.Top, candidate.Bounds.Bottom))
            .OrderBy(candidate => DistanceToOverlap(
                crossingY,
                Math.Max(current.Top, candidate.Bounds.Top),
                Math.Min(current.Bottom, candidate.Bounds.Bottom)))
            .ThenBy(candidate => candidate.Number)
            .FirstOrDefault();
    }

    private DisplayMonitor? FindVerticalNeighbor(
        int direction,
        double crossingX)
    {
        var current = Display.Bounds;
        return _displays
            .Where(candidate => candidate.DeviceName != Display.DeviceName)
            .Where(candidate =>
                direction > 0
                    ? candidate.Bounds.Top == current.Bottom
                    : candidate.Bounds.Bottom == current.Top)
            .Where(candidate =>
                Overlaps(current.Left, current.Right,
                    candidate.Bounds.Left, candidate.Bounds.Right))
            .OrderBy(candidate => DistanceToOverlap(
                crossingX,
                Math.Max(current.Left, candidate.Bounds.Left),
                Math.Min(current.Right, candidate.Bounds.Right)))
            .ThenBy(candidate => candidate.Number)
            .FirstOrDefault();
    }

    private static bool Overlaps(
        int firstStart,
        int firstEnd,
        int secondStart,
        int secondEnd) =>
        Math.Max(firstStart, secondStart) < Math.Min(firstEnd, secondEnd);

    private static double DistanceToOverlap(
        double position,
        int overlapStart,
        int overlapEnd)
    {
        var lastPixel = overlapEnd - 1;
        if (position < overlapStart)
        {
            return overlapStart - position;
        }

        return position > lastPixel ? position - lastPixel : 0;
    }

    private DisplayMonitor FindDisplay(Point point)
    {
        return _displays.FirstOrDefault(
                display => display.Bounds.Contains(point))
            ?? _displays
                .OrderBy(display =>
                {
                    var x = Math.Clamp(
                        point.X,
                        display.Bounds.Left,
                        display.Bounds.Right - 1);
                    var y = Math.Clamp(
                        point.Y,
                        display.Bounds.Top,
                        display.Bounds.Bottom - 1);
                    var dx = point.X - x;
                    var dy = point.Y - y;
                    return (long)dx * dx + (long)dy * dy;
                })
                .First();
    }

    private Point RoundedPosition()
    {
        var bounds = Display.Bounds;
        return new Point(
            Math.Clamp(
                (int)Math.Round(_mapX),
                bounds.Left,
                bounds.Right - 1),
            Math.Clamp(
                (int)Math.Round(_mapY),
                bounds.Top,
                bounds.Bottom - 1));
    }

    public static IReadOnlyList<DisplayMonitor> GetDisplays()
    {
        return Forms.Screen.AllScreens
            .OrderBy(screen => screen.Bounds.Left)
            .ThenBy(screen => screen.Bounds.Top)
            .Select((screen, index) => new DisplayMonitor(
                screen.DeviceName,
                screen.Bounds,
                screen.WorkingArea,
                screen.Primary,
                index + 1))
            .ToList();
    }
}

File: linux/precision_jump/screen_targeting.py
"""Port of Services/ScreenTargeting.cs (NineGridSession).

This is the heart of Precision Jump: it maps physical mouse displacement onto
a zoomable model of the display topology.

  depth 0   -> the coarsest "screen map": displays are connected through a
               simplified eight-direction topology, so leaving one display's
               normalised [0,1] box hops to the nearest display in that octant.
  depth > 0 -> the "actual layout map": movement uses the real X screen
               geometry (positions, sizes, alignment and gaps).

Scale is 3**depth and is eased over ScaleAnimationSeconds. The scale never
moves the pointer by itself -- it only changes how much map distance a given
physical displacement buys.

Timestamps are seconds (time.perf_counter) rather than Stopwatch ticks; every
public method accepts an explicit `now` so the logic stays deterministic under
test.
"""

from __future__ import annotations

import math
import time
from dataclasses import dataclass, replace
from typing import Iterable, Optional, Sequence

from .geometry import DisplayMonitor, Point, Rectangle

SCALE_ANIMATION_SECONDS = 0.14
_EPSILON = 0.000001


@dataclass(frozen=True)
class NineGridPreview:
    display: DisplayMonitor
    actual_cursor: Point
    depth: int
    map_scale: float
    step_x: float
    step_y: float

    @property
    def is_screen_level(self) -> bool:
        return self.depth == 0

    @property
    def label(self) -> str:
        if self.is_screen_level:
            return f"SCREEN {self.display.number}"
        return f"ZOOM {self.map_scale:.1f}×"

    def with_display(self, display: DisplayMonitor, **changes) -> "NineGridPreview":
        return replace(self, display=display, **changes)


@dataclass(frozen=True)
class GridInteraction:
    preview: NineGridPreview
    jump_target: Optional[Point]


def _clamp(value, low, high):
    return max(low, min(high, value))


def _contains(bounds: Rectangle, x: float, y: float) -> bool:
    return (
        x >= bounds.left
        and x < bounds.right
        and y >= bounds.top
        and y < bounds.bottom
    )


def _octant(dx: float, dy: float) -> tuple[int, int]:
    """Quantise a direction into one of eight (row, column) steps."""
    angle = math.atan2(dy, dx)
    raw = angle / (math.pi / 4)
    # C# MidpointRounding.AwayFromZero (Python's round() is banker's rounding).
    octant = int(math.floor(raw + 0.5)) if raw >= 0 else int(math.ceil(raw - 0.5))
    octant = ((octant % 8) + 8) % 8
    return {
        0: (0, 1),
        1: (1, 1),
        2: (1, 0),
        3: (1, -1),
        4: (0, -1),
        5: (-1, -1),
        6: (-1, 0),
        7: (-1, 1),
    }[octant]


def _update_ray_interval(origin, direction, min_bound, max_bound, interval):
    """Slab clip for one axis. `interval` is a mutable [minimum, maximum]."""
    if abs(direction) < _EPSILON:
        return min_bound <= origin <= max_bound

    first = (min_bound - origin) / direction
    second = (max_bound - origin) / direction
    if first > second:
        first, second = second, first

    interval[0] = max(interval[0], first)
    interval[1] = min(interval[1], second)
    return interval[1] >= interval[0]


def _ray_rectangle_entry(
    origin_x, origin_y, direction_x, direction_y, rectangle: Rectangle
) -> Optional[float]:
    """Parametric distance at which a ray enters a rectangle, or None."""
    interval = [-math.inf, math.inf]
    if not _update_ray_interval(
        origin_x, direction_x, rectangle.left, rectangle.right, interval
    ) or not _update_ray_interval(
        origin_y, direction_y, rectangle.top, rectangle.bottom, interval
    ):
        return None

    entry = max(interval[0], 0.0)
    return entry if interval[1] >= entry else None


class NineGridSession:
    def __init__(
        self,
        origin: Point,
        displays: Sequence[DisplayMonitor],
    ) -> None:
        if not displays:
            raise RuntimeError("No displays are available.")

        self._displays = list(displays)
        self.origin = origin
        self._map_x = float(origin.x)
        self._map_y = float(origin.y)
        self._depth = 0
        self._scale = 1.0
        self._scale_start = 1.0
        self._scale_target = 1.0
        self._scale_started_at = 0.0
        self.display = self._find_display(origin)
        self.preview = self._create_preview()

    # ---------------------------------------------------------------- state

    @property
    def current_position(self) -> Point:
        return self._rounded_position()

    @property
    def depth(self) -> int:
        return self._depth

    @property
    def map_scale(self) -> float:
        return self._scale

    @property
    def step_x(self) -> float:
        return self.display.bounds.width / max(self._scale, 1.0)

    @property
    def step_y(self) -> float:
        return self.display.bounds.height / max(self._scale, 1.0)

    def is_scale_animating(self, now: Optional[float] = None) -> bool:
        now = time.perf_counter() if now is None else now
        elapsed = now - self._scale_started_at
        return (
            abs(self._scale - self._scale_target) > 0.001
            and elapsed < SCALE_ANIMATION_SECONDS
        )

    # ------------------------------------------------------------- movement

    def move_continuous(
        self,
        physical_delta_x: float,
        physical_delta_y: float,
        map_unit_travel_distance: float,
        now: Optional[float] = None,
    ) -> GridInteraction:
        self._update_scale(now)
        if physical_delta_x == 0 and physical_delta_y == 0:
            self.preview = self._create_preview()
            return GridInteraction(self.preview, None)

        travel = max(map_unit_travel_distance, 1.0)
        if self._depth == 0:
            self._move_on_discrete_screen_map(
                physical_delta_x / (travel * self._scale),
                physical_delta_y / (travel * self._scale),
            )
        else:
            self._move_on_actual_display_map(
                physical_delta_x * self.display.bounds.width / (travel * self._scale),
                physical_delta_y * self.display.bounds.height / (travel * self._scale),
            )

        self.preview = self._create_preview()
        return GridInteraction(self.preview, self.current_position)

    def zoom_in(
        self,
        maximum_depth: int,
        actual_position: Optional[Point] = None,
        now: Optional[float] = None,
    ) -> GridInteraction:
        if self._depth >= maximum_depth:
            return GridInteraction(self.preview, None)

        if actual_position is not None:
            self._sync_position(actual_position)
        self._start_scale_transition(self._depth + 1, now)
        self.preview = self._create_preview()
        return GridInteraction(self.preview, None)

    def zoom_out(
        self,
        actual_position: Optional[Point] = None,
        now: Optional[float] = None,
    ) -> GridInteraction:
        if self._depth == 0:
            return GridInteraction(self.preview, None)

        if actual_position is not None:
            self._sync_position(actual_position)
        self._start_scale_transition(self._depth - 1, now)
        self.preview = self._create_preview()
        return GridInteraction(self.preview, None)

    def refresh_scale(self, now: Optional[float] = None) -> NineGridPreview:
        self._update_scale(now)
        self.preview = self._create_preview()
        return self.preview

    def accept_programmatic_position(self, actual_position: Point) -> None:
        self._sync_position(actual_position)
        self.preview = self._create_preview()

    # ------------------------------------------------------- internal moves

    def _move_on_discrete_screen_map(self, normalized_dx, normalized_dy) -> None:
        bounds = self.display.bounds
        width = max(bounds.width - 1, 1)
        height = max(bounds.height - 1, 1)
        u = (self._map_x - bounds.left) / width + normalized_dx
        v = (self._map_y - bounds.top) / height + normalized_dy

        for _ in range(8):
            column_direction = -1 if u < 0 else (1 if u > 1 else 0)
            row_direction = -1 if v < 0 else (1 if v > 1 else 0)
            if row_direction == 0 and column_direction == 0:
                break

            next_display = self._find_discrete_display(row_direction, column_direction)
            if next_display is None:
                u = _clamp(u, 0, 1)
                v = _clamp(v, 0, 1)
                break

            if column_direction > 0:
                u -= 1
            elif column_direction < 0:
                u += 1
            if row_direction > 0:
                v -= 1
            elif row_direction < 0:
                v += 1

            self.display = next_display

        bounds = self.display.bounds
        self._map_x = bounds.left + _clamp(u, 0, 1) * max(bounds.width - 1, 1)
        self._map_y = bounds.top + _clamp(v, 0, 1) * max(bounds.height - 1, 1)

    def _move_on_actual_display_map(self, map_delta_x, map_delta_y) -> None:
        proposed_x = self._map_x + map_delta_x
        proposed_y = self._map_y + map_delta_y
        if _contains(self.display.bounds, proposed_x, proposed_y):
            self._map_x = proposed_x
            self._map_y = proposed_y
            return

        containing = next(
            (d for d in self._displays if _contains(d.bounds, proposed_x, proposed_y)),
            None,
        )
        if containing is not None:
            self.display = containing
            self._map_x = proposed_x
            self._map_y = proposed_y
            return

        crossing = self._find_actual_layout_crossing(
            self._map_x, self._map_y, map_delta_x, map_delta_y
        )
        if crossing is not None:
            target_display, entry_x, entry_y = crossing
            self.display = target_display
            self._map_x = entry_x
            self._map_y = entry_y
            return

        # Keep integrating through real virtual-desktop gaps. The visible
        # pointer stays projected to the current display edge until the
        # continuous map coordinate enters another display rectangle.
        self._map_x = proposed_x
        self._map_y = proposed_y

    # ------------------------------------------------------------ zoom math

    def _start_scale_transition(self, depth: int, now: Optional[float]) -> None:
        now = time.perf_counter() if now is None else now
        self._update_scale(now)
        self._depth = depth
        self._scale_start = self._scale
        self._scale_target = math.pow(3, depth)
        self._scale_started_at = now

    def _update_scale(self, now: Optional[float]) -> None:
        if abs(self._scale - self._scale_target) <= 0.001:
            self._scale = self._scale_target
            return

        now = time.perf_counter() if now is None else now
        elapsed = now - self._scale_started_at
        progress = _clamp(elapsed / SCALE_ANIMATION_SECONDS, 0.0, 1.0)
        eased = progress * progress * (3 - 2 * progress)  # smoothstep
        self._scale = self._scale_start + (self._scale_target - self._scale_start) * eased
        if progress >= 1:
            self._scale = self._scale_target

    # -------------------------------------------------------------- lookups

    def _sync_position(self, position: Point) -> None:
        self._map_x = float(position.x)
        self._map_y = float(position.y)
        self.display = self._find_display(position)

    def _create_preview(self) -> NineGridPreview:
        return NineGridPreview(
            self.display,
            self.current_position,
            self._depth,
            self._scale,
            self.step_x,
            self.step_y,
        )

    def _find_discrete_display(
        self, row_direction: int, column_direction: int
    ) -> Optional[DisplayMonitor]:
        current_center = self.display.bounds.center
        candidates = []
        for candidate in self._displays:
            if candidate.device_name == self.display.device_name:
                continue
            candidate_center = candidate.bounds.center
            dx = candidate_center.x - current_center.x
            dy = candidate_center.y - current_center.y
            if _octant(dx, dy) != (row_direction, column_direction):
                continue
            candidates.append((dx * dx + dy * dy, candidate))

        if not candidates:
            return None
        # min() on the distance only -- DisplayMonitor is not orderable and
        # ties must keep enumeration order, like C#'s stable OrderBy.
        return min(candidates, key=lambda item: item[0])[1]

    def _find_actual_layout_crossing(
        self, origin_x, origin_y, direction_x, direction_y
    ):
        length = math.sqrt(direction_x * direction_x + direction_y * direction_y)
        if length <= 0:
            return None

        best = None
        for candidate in self._displays:
            if candidate.device_name == self.display.device_name:
                continue
            entry = _ray_rectangle_entry(
                origin_x, origin_y, direction_x, direction_y, candidate.bounds
            )
            if entry is None or not (0 <= entry <= 1):
                continue
            if best is None or entry < best[0]:
                best = (entry, candidate)

        if best is None:
            return None

        entry, target = best
        unit_x = direction_x / length
        unit_y = direction_y / length
        bounds = target.bounds
        return (
            target,
            _clamp(origin_x + direction_x * entry + unit_x, bounds.left, bounds.right - 1),
            _clamp(origin_y + direction_y * entry + unit_y, bounds.top, bounds.bottom - 1),
        )

    def _find_display(self, point: Point) -> DisplayMonitor:
        for display in self._displays:
            if display.bounds.contains(point):
                return display

        def squared_gap(display: DisplayMonitor) -> int:
            x = _clamp(point.x, display.bounds.left, display.bounds.right - 1)
            y = _clamp(point.y, display.bounds.top, display.bounds.bottom - 1)
            dx = point.x - x
            dy = point.y - y
            return dx * dx + dy * dy

        return min(self._displays, key=squared_gap)

    def _rounded_position(self) -> Point:
        bounds = self.display.bounds
        return Point(
            int(_clamp(_round_half_away(self._map_x), bounds.left, bounds.right - 1)),
            int(_clamp(_round_half_away(self._map_y), bounds.top, bounds.bottom - 1)),
        )


def _round_half_away(value: float) -> int:
    """Math.Round(value) in C# rounds .5 away from zero; Python rounds to even."""
    return int(math.floor(value + 0.5)) if value >= 0 else int(math.ceil(value - 0.5))


def find_grid_cell(bounds: Rectangle, grid_size: int, point: Point) -> tuple[int, int]:
    """Helper retained for parity with the C# test-suite entry points."""
    cell_width = bounds.width / grid_size
    cell_height = bounds.height / grid_size
    column = int(_clamp(math.floor((point.x - bounds.left) / cell_width), 0, grid_size - 1))
    row = int(_clamp(math.floor((point.y - bounds.top) / cell_height), 0, grid_size - 1))
    return row, column


def grid_cell(bounds: Rectangle, grid_size: int, row: int, column: int) -> Rectangle:
    cell_width = bounds.width / grid_size
    cell_height = bounds.height / grid_size
    left = int(round(bounds.left + column * cell_width))
    top = int(round(bounds.top + row * cell_height))
    return Rectangle(left, top, int(round(cell_width)), int(round(cell_height)))


def order_displays(displays: Iterable[DisplayMonitor]) -> list[DisplayMonitor]:
    """Match GetDisplays(): order by left, then top, numbering from 1."""
    ordered = sorted(displays, key=lambda d: (d.bounds.left, d.bounds.top))
    return [
        DisplayMonitor(
            d.device_name, d.bounds, d.working_area, d.is_primary, index + 1
        )
        for index, d in enumerate(ordered)
    ]
